homeassistant/core.py

```diff
@@ -845,15 +846,21 @@ def async_fire(
         if not listeners:
             return
 
-        for job, event_filter in listeners:
+        for job, event_filter, run_immediately in listeners:
             if event_filter is not None:
                 try:
                     if not event_filter(event):
                         continue
                 except Exception:  # pylint: disable=broad-except
                     _LOGGER.exception("Error in event filter")
                     continue
-            self._hass.async_add_hass_job(job, event)
+            if run_immediately:
+                try:
+                    job.target(event)
+                except Exception:  # pylint: disable=broad-except
+                    _LOGGER.exception("Error running job: %s", job)
```

Comment:
I was going to suggest:
```suggestion
                self._hass.async_run_hass_job(job, event)
```
but then realized that it doesn't catch exceptions.